---
title: "The HCABrowser Package"
package: HCABrowser
author:
- name: "Daniel Van Twisk"
- name: "Martin Morgan"
  affiliation: "Roswell Park Comprehensive Cancer Center"
output:
  BiocStyle::html_document
abstract: >
  The Human Cell Atlas (HCA) (https://www.humancellatlas.org/) was created in
  order to create comprehensive reference maps of all human cells as a basis for
  both understanding human health and diagnosing, monitoring, and treating
  disease. The *HCABrowser* Biocondctor pacakge provides infrastructure for
  searching for, queerying, and accessing data help on the Human Cell Atlas's
  Data Coordination Platform (https://dss.data.humancellatlas.org/). Further
  changes to the package are planned to incorperate higer level functionality to
  upload user generated data to the the Human Cell Atlas platform.
vignette: >
  %\VignetteIndexEntry{The HCABrowser Package}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r init, results='hide', echo=FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE)
```

# What is the Human Cell Atlas?

From the [Human Cell Atlas (HCA) website](https://www.humancellatlas.org/):

> The cell is the core unit of the human body -- the key to
understanding the biology of health and the ways in which molecular
dysfunction leads to disease.  Yet our characterization of the
hundreds of types and subtypes of cells in the human body is limited,
based partly on techniques that have limited resolution and
classifications that do not always map neatly to each other. Genomics
has offered a systematic approach, but it has largely been applied in
bulk to many cell types at once -- masking critical differences
between cells -- and in isolation from other valuable sources of data.

> Recent advances in single-cell genomic analysis of cells and tissues have put
systematic, high-resolution and comprehensive reference maps of all human cells
within reach. In other words, we can now realistically envision a human cell
atlas to serve as a basis for both understanding human health and diagnosing,
monitoring, and treating disease.

> At its core, a cell atlas would be a collection of cellular reference maps,
characterizing each of the thousands of cell types in the human body and where
they are found. It would be an extremely valuable resource to empower the global
research community to systematically study the biological changes associated
with different diseases, understand where genes associated with disease are
active in our bodies, analyze the molecular  mechanisms that govern the
production and activity of different cell types, and sort out how different cell
types combine and work together to form tissues.

The Human Cell Atlas facilitates queries on it's [data coordination platform with
a RESTFUL API] (https://dss.data.humancellatlas.org/).

## Installation

To install this package, use Bioconductor's `BiocManager` package.

```{r install_bioc, eval=FALSE}
if (!require("BiocManager"))
    install.packages("BiocManager")
BiocManager::install('HCABrowser')
```

```{r libraries, message=FALSE}
library(HCABrowser)
```

## Connecting to the Human Cell Atlas

The [HCABrowser][] package relies on having network
connectivety. Also, the HCA's Data Coordination Platform (DCP) must
also be operational. This package is meant for users who have a
working knowledge of the HCA DCP's schema to provide the user with the
fill richness of the HCA's content. For a more simple-to-use way to
access and download data from The Human Cell Atlas, please use the
`HCAExplorer` package.

[HCABrowser]: https://bioconductor.org/packages/HCABrowser

The `HCABrowser` object serves as the representation of the Human Cell
Atlas. Upon creation, it will automatically peform a cursorary query and
display a small table showing the first few bundles of the entire HCA. This
intial table contains some columns that we have determined are most useful
to users. The output also displays the url of the instance of the HCA DCP being
used, the current query, whether bundles or files are being displayed, and the
number of bundles in the results

By default, ten bundles per page will be displayed in the result for "raw"
output and 100 will be displayed for "summary" output. The `host` argument
dictates the hca-dcp the client will be accessing while the `api_url` argument
dictates what schema the client will be accessing.  These three values can be
changed in the constructor. The default values are given below.

If the HCA cannot be reached, an error will be thrown displaying the status of
the request.

```{r createHCA}
hca <-
    HCABrowser(
        api_url='https://dss.data.humancellatlas.org/v1/swagger.json',
        host = 'dss.data.humancellatlas.org/v1'
    )
hca
```

Upon displaying the object, multiple fields can be seen:
- The class: `HCABrowser`
- The hca dcp host address that is currently being used.
- The current query (assigned using `filter()`)
- The current selection (assigned using `select()`)
    - You may notice that some columns are already selected. These columns are
      automatically selected to allow the user some initial view of the hca.

To change how many pages are being displayed, the `per_page()` method can be
used.
(Note the hca dcp had a maximum of 10 bundles per page to be shown at a time)

Since there are far more bundles in the HCA than can be shown, if `link` is
`True`, the next set of bundles can be obtained using the `nextResults` method.

```{r nextResults, eval=FALSE}
hca <- nextResults(hca)
hca
```

## Querying the HCABrowser

The HCA extends the functionality of the [dplyr][] package's
`filter()` and `select()` methods.

[dplyr]: https://cran.r-project.org/package=dplyr

The `filter()` method allows the user to query the HCA by relating fields to
certain values. Character fields can be queried using the operators:
- `==`
- `!=`
- `%in%`
- `%startsWith%`
- `%endsWith%`
- `%contains%`

Numeric fields can be queried with the operators:
- `==`
- `!=`
- `%in%`
- `>`
- `<`
- `>=`
- `<=`

Queries can be encompassed by parenthesese
- `()`

Queries can be negated by placing the `!` symbol in front

Combination operators can be used to combine queries
- `&`
- `|`

Now we see that "brain" and "Brain" are available values. Since these values are
the result of input by other users, there may be errors or inconsistencies. To
be safe, both fields can be queried with the following query:

```{r firstFilter}
hca2 <-
    hca %>%
    filter('files.specimen_from_organism_json.organ.text' == c('Brain', 'brain'))
hca2 <-
    hca %>%
    filter(
        'files.specimen_from_organism_json.organ.text' %in% c('Brain', 'brain')
    )
hca2 <-
    hca %>%
    filter(
        'files.specimen_from_organism_json.organ.text' == Brain |
        'files.specimen_from_organism_json.organ.text' == brain
    )
hca2
```

If we also wish to search for results based on the NCBI Taxon ID for mouse,
10090, as well as brain, we can perform this query in a variety of ways.
```{r multiFilter}
hca2 <-
    hca %>%
    filter(
        'files.specimen_from_organism_json.organ.text' %in% c('Brain', 'brain')
    ) %>%
    filter('specimen_from_organism_json.biomaterial_core.ncbi_taxon_id' == 10090)
hca2 <-
    hca %>%
    filter(
        'files.specimen_from_organism_json.organ.text' %in% c('Brain', 'brain'),
        'specimen_from_organism_json.biomaterial_core.ncbi_taxon_id' == 10090
    )
hca <-
    hca %>%
    filter(
        'files.specimen_from_organism_json.organ.text' %in% c('Brain', 'brain') &
        'specimen_from_organism_json.biomaterial_core.ncbi_taxon_id' == 10090
    )
hca
```

The `HCABrowser` package is able to handle arbitrarily complex queries on the
Human Cell Atlas.

```{r complexFilter}
hca2 <-
    hca %>%
    filter((!organ.text %in% c('Brain', 'blood')) & (
        files.specimen_from_organism_json.genus_species.text == "Homo sapiens" |
        library_preparation_protocol_json.library_construction_approach.text == 'Smart-seq2'
    ))
hca2
``` 

The `HCABrowser` object can undo the most recent queries run on it.

```{r undoQuery}
hca <-
    hca %>%
    filter('files.specimen_from_organism_json.organ.text' == heart)
hca <-
    hca %>%
    filter('files.specimen_from_organism_json.organ.text' != brain)
hca <- undoEsQuery(hca, n = 2)
hca
```

If one would want to start from a fresh query but retain the modifications made
to the `HCABrowser` object, the `resetEsQuery()` method can be used.

```
hca <- resetEsQuery(hca)
hca
```

Using `fields()`, we can find that the fields `paired_end` and
`organ.ontology` are availiable. These fields can be shown in our resulting
`HCABrowser` object using the `select()` method.

```{r select}
hca2 <-
    hca %>%
    select('paired_end', 'organ.ontology')
hca2 <-
    hca %>%
    select(c('paired_end', 'organ.ontology'))
hca2
```

# Use examples

The remainder of this vignette will review applicable uses for this package.

## Download a Bundle

The following retrieves the first ten results satisfying the filter
criterion.

```{r getBundle}
hca <-
    HCABrowser() %>%
    filter(
        'files.specimen_from_organism_json.organ.text' == "brain"
    )
bundles <- searchBundles(hca, replica='aws', output_format='raw')
bundles %>% httr::content() %>% names()
result <- parseToSearchResults(bundles)
```

Each result is represented as a 'bundle' of primary data and
associated metadata. Bundles have a unique identifier, `bundle_fqid`,
consisting of an identifier and version (date) stamp)

```{r}
fqids <- vapply(results(result), `[[`, character(1), "bundle_fqid")
uuid <- sub("([[:alnum:]-]+)\\.(.*)", "\\1", fqids)
version <- sub("([[:alnum:]-]+)\\.(.*)", "\\2", fqids)
tibble(uuid, version)
```

Use `getBundle()` to 'check out' a bundle from the HCA to the Amazon
or Google cloud.

```{r}
## provide 'version' for precise control
bundle <- getBundle(hca, uuid[1], replica = "aws")
bundle %>%
    str()
```

The `checkout_job_id` can be used to query the status of the bundle

```{r}
jid <- as.list(bundle)$checkout_job_id
status <- getBundleCheckout(hca, checkout_job_id=jid, replica='aws')
status %>% str()
```

The `status` will change from 'RUNNING' to 'SUCCEEDED', showing the
location of the bundle for download via AWS or GCP command line tools.

```
> getBundleCheckout(hca, checkout_job_id=jid, replica='aws') %>% str()
List of 1
 $ status: chr "RUNNING"
> ## time passes
> getBundleCheckout(hca, checkout_job_id=jid, replica='aws') %>% str()
List of 2
 $ location: chr "s3://org-hca-dss-checkout-prod/bundles/fffcea5e-...
 $ status  : chr "SUCCEEDED"
```

## Download a File

```{r getFile}
hca <-
    HCABrowser() %>%
    filter('files.specimen_from_organism_json.organ.text' == "brain")
results <-
    searchBundles(hca, replica='aws', output_format='raw') %>%
    parseToSearchResults() %>%
    results()
result <- results[[1]]
```

`result` is a list-of-lists structure. Files associated with a bundle
are in at the following path:

```{r}
tbl <-
    result$metadata$manifest$files %>%
    bind_rows() %>%
    mutate(`content-type` = noquote(`content-type`))
tbl
```

The following filters the manifest to the name and uuid of a fastq
file

```{r}
fastq <- 
    tbl %>%
    filter(endsWith(name, "fastq.gz")) %>%
    select(name, uuid) %>%
    head(1)
fastq
```

Download the file with

```{r, eval = FALSE}
getFile(HCABrowser(), uuid=fastq$uuid, replica='aws')
```

# Developer notes

- The `S3` object-oriented programming paradigm is used.
- Methods from the `dplyr` package can be used to manipulate objects in the
`HCABrowser` package.
- In the future, we wish to expand the functionalit of this packages to cover
the remaining functionality of the hca dcp api.

# Session info {.unnumbered}

```{r sessionInfo, echo = FALSE}
sessionInfo()
```
